cmd/pitr/collector/collector.go

```diff
@@ -0,0 +1,264 @@
+	}
+	lastSet, err := ioutil.ReadAll(lastSetObject)
+	if err != nil && !strings.Contains(err.Error(), "The specified key does not exist") {
+		return nil, errors.Wrap(err, "read object")
+	}
+
+	lastSet = []byte("")
+	return &Collector{
+		storage:        s3,
+		lastSet:        string(lastSet),
+		pxcUser:        c.PXCUser,
+		pxcPass:        c.PXCPass,
+		pxcServiceName: c.PXCServiceName,
+	}, nil
+}
+
+func (c *Collector) Run() error {
+	err := c.newDB()
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files")
+	}
+
+	return nil
+}
+
+func (c *Collector) newDB() error {
+	host, err := c.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Collector) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Collector) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+c.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ":")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost, nil
+}
+
+func (c *Collector) CollectBinLogs() error {
+	// get last uploaded binlog file name
+	binlogName, err := c.db.GetBinLogName(c.lastSet)
+	if err != nil {
+		return errors.Wrap(err, "get binlog name by set")
+	}
+	list, err := c.db.GetBinLogList()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+
+	upload := false
+	// if there are no uploaded files we going to upload every binlog file
+	if len(binlogName) == 0 {
+		upload = true
+	}
+
+	for _, binlog := range list {
+		if binlog == binlogName { // this check is for uploading starting from needed file
+			upload = true
+		}
+		if upload {
+			err = c.manageBinlog(binlog)
+			if err != nil {
+				return errors.Wrap(err, "manage binlog")
+			}
+		}
+	}
+
+	return nil
+}
+
+type reader struct {
+	r io.Reader
+}
+
+func (r *reader) Read(p []byte) (int, error) {
+	return r.r.Read(p)
+}
+
+func (c *Collector) manageBinlog(binlog string) (err error) {
+	set, err := c.db.GetGTIDSet(binlog)
+	if err != nil {
+		return errors.Wrap(err, "get GTID set")
+	}
+	if len(set) == 0 {
+		return nil
+	}
+
+	binlogTmstmp, err := c.db.GetBinLogFirstTimestamp(binlog)
+	if err != nil {
+		return errors.Wrapf(err, "get first timestamp for %s", binlog)
+	}
+
+	binlogName := "binlog:" + binlogTmstmp
+
```

Comment:
binlog name should be uniq, lets add md5 of gtidSet to filename
```suggestion
	binlogName := "binlog_" + binlogTmstmp + "_" + md5(gtidSet)
```